master/internal/stream/supervisor.go

```diff
@@ -0,0 +1,150 @@
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/pkg/syncx/errgroupx"
+)
+
+const (
+	minInterval = time.Second
+	maxInterval = time.Minute
+	maxRetries  = 5
+)
+
+// Supervisor manages the context for underlying PublisherSet.
+type Supervisor struct {
+	cond *sync.Cond
+	// these things change each restart of the PublisherSet
+	subctx context.Context
+	ps     *PublisherSet
+}
+
+// NewSupervisor creates a new Supervisor.
+func NewSupervisor() *Supervisor {
+	var lock sync.Mutex
+	return &Supervisor{cond: sync.NewCond(&lock)}
+}
+
+// monitorPermissionChanges listens for permission changes, updates the PublisherSet
+// to signal to boot streamers, returns an error in the event of a failure to listen.
+func monitorPermissionChanges(ctx context.Context, ps *PublisherSet) error {
+	permListener, err := AuthZProvider.Get().GetPermissionChangeListener()
+	if err != nil {
+		log.Errorf("unable to get permission change listener: %s", err)
+		// XXX: investigate better recovery mechanism
+		return err
+	}
+	if permListener == nil {
+		// no listener means we don't have permissions configured at all
+		return nil
+	}
+	defer func() {
+		err := permListener.Close()
+		if err != nil {
+			log.Debugf("error occurred while closing permission listener: %s", err)
+		}
+	}()
+
+	for {
+		select {
+		// did permissions change?
+		case <-permListener.Notify:
+			log.Debugf("permission change detected, booting streamers")
+			func() {
+				ps.permissionLock.Lock()
+				defer ps.permissionLock.Unlock()
+				close(ps.permissionChangeChan)
+				ps.permissionChangeChan = make(chan struct{})
+			}()
+		// is the listener still alive?
+		case <-time.After(30 * time.Second):
+			pingErrChan := make(chan error)
+			go func() {
+				err = permListener.Ping()
+				pingErrChan <- errors.Wrap(err, "no active connection")
+			}()
+			if err := <-pingErrChan; err != nil {
+				log.Errorf("permission listener failed %s", err)
+				return err
+			}
+		// are we canceled?
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (ssup *Supervisor) runOne(ctx context.Context) error {
+	group := errgroupx.WithContext(ctx)
+	subctx := group.Context()
+	func() {
+		ssup.cond.L.Lock()
+		defer ssup.cond.L.Unlock()
+		ssup.cond.Broadcast()
+
+		ssup.subctx = subctx
+		ssup.ps = NewPublisherSet()
+
+		// start monitoring permissions
+		group.Go(
+			func(c context.Context) error {
+				return monitorPermissionChanges(c, ssup.ps)
+			},
+		)
+		// start up all publishers
+		// XXX: should these be passing errors backup?
+		group.Go(ssup.ps.Start)
+		// group.Go(PublishLoop(ctx, "experiment_trial_chan", ps.Experiments))
+	}()
+	return group.Wait()
+}
+
+// Run attempts to start up the publisher system and recovers in the event of a failure.
+func (ssup *Supervisor) Run(ctx context.Context) error {
+	backoffSettings := backoff.NewExponentialBackOff()
+	backoffSettings.InitialInterval = minInterval
+	backoffSettings.MaxInterval = maxInterval
```

Comment:
Something isn't quite right here.

We know nobody's actually checking ssup.Run for errors.  So why not retry forever?

Or put a `for {}` loop in the top of the core.api goroutine that calls Run.

If you put a `for {}` outside ssup.Run() then your exponential backoff will keep resetting, in permafail scenarios, which is weird.

But if you just grant infinite retries to this ExponentialBackoff, then you'll effectively reset faster during a network blip today than you do tomorrow or next week.  Which is weird.

I dunno, like each component makes sense by itself the way you have it written but when you look at the whole picture (with no top-level supervisor) something feels off.

I don't have a particular recommendation for how to proceed, and honestly if you shipped exactly this and it broke in production you could reasonably say, "well there's no top-level supervisor and that's way out of scope for streaming updates, not my fault".

Reply:
I don't think adding a top-level supervisor is in scope for this stage of streaming updates by any means, that really should be it's own unit of work since it'll touch everything in `core.go:Run()`, and likely require refactor of everything that is assuming there is no top-level supervisor.

You do bring up a good point, that this could cause some unexpected delays & loss of the `stream.Supervisor` that would cause the backing function for `/stream` to also be invalid.

I could either put the infinite loop in to remain consistent with the current system or replace the exponential backoff with a constant backoff w/unlimited retries (effectively the same thing).

I'm leaning towards the `for {}` since what's going on would be clearer and it would be clear that it needs to be refactored in the future.